### PR TITLE
Extract performance improvements

### PR DESCRIPTION
**Pull request by dev7355608**
##### Description of change

- 3d2d89c9d6ad1a6fe6bfb78fec5a2a4dc10d9d8f: `canvas.toBlob` instead of `toDataURL` blocks the main thread for a shorter duration.
- 4a2ff59114c9f015b9c96f7f1c04fd4e53c331c2: Improves performance (~50ms => ~42ms in my test). The `getImageData` is unnessecary.
- 4b2ce461ddd34465d27a138ed245c7eaf9d876ca: Improves performance even more (~42ms => ~34ms in my test).

##### Pre-Merge Checklist

- [x] Lint process passed (`npm run lint`)
- [x] Tests passed (`npm run test`)

### CODE REVIEW

File: packages/canvas-extract/src/CanvasExtract.ts [hunk 1]
```diff
@@ -63,6 +63,20 @@ export class CanvasExtract implements ISystem, IExtract
     {
         const canvas = this.canvas(target);
 
+        if (canvas.toBlob !== undefined)
+        {
+            return await new Promise<string>((resolve, reject) =>
```
Comment (bigtimebuddy) [inline packages/canvas-extract/src/CanvasExtract.ts:68]: The await here is unnecessary and creates an additional promise. We should remove it.

```suggestion
            return new Promise<string>((resolve, reject) =>
```

File: packages/canvas-extract/src/CanvasExtract.ts [hunk 2]
```diff
@@ -71,16 +85,18 @@ export class CanvasExtract implements ISystem, IExtract
         {
             const blob = await canvas.convertToBlob({ type: format, quality });
 
-            return await new Promise<string>((resolve) =>
+            return await new Promise<string>((resolve, reject) =>
```
Comment (bigtimebuddy) [inline packages/canvas-extract/src/CanvasExtract.ts:88]: Same here. 
```suggestion
            return new Promise<string>((resolve, reject) =>
```

File: packages/extract/src/Extract.ts [hunk 3]
```diff
@@ -84,6 +84,20 @@ export class Extract implements ISystem, IExtract
     {
         const canvas = this.canvas(target);
 
+        if (canvas.toBlob !== undefined)
+        {
+            return await new Promise<string>((resolve, reject) =>
```
Comment (bigtimebuddy) [inline packages/extract/src/Extract.ts:89]:
```suggestion
            return new Promise<string>((resolve, reject) =>
```

File: packages/extract/src/Extract.ts [hunk 4]
```diff
@@ -92,16 +106,18 @@ export class Extract implements ISystem, IExtract
         {
             const blob = await canvas.convertToBlob({ type: format, quality });
 
-            return await new Promise<string>((resolve) =>
+            return await new Promise<string>((resolve, reject) =>
```
Comment (bigtimebuddy) [inline packages/extract/src/Extract.ts:109]:
```suggestion
            return new Promise<string>((resolve, reject) =>
```

**Review by bigtimebuddy — APPROVED** [review]
Nice one!

**Comment by dev7355608** [comment]
I noticed my y-flipping was flawed: I forgot about the 4 bytes per pixel. There was no test that could have caught this error: so I added a bunch of tests that checks that looks that the extracted pixels/canvas.

I also replaced `postArrayDivide` with a much faster version (more that 3 times faster in my test).

**Comment by dev7355608** [comment]
I added test for `CanvasExtract` too. Notice that `Extract.pixels` from a renderer does not y-flip, while `CanvasExtract.pixels` returns y-flipped results. Is this a bug? Wouldn't it make more sense if `Extract.pixels` y-flipped?

Another difference that bothers me a little bit is that `CanvasExtract.pixels` returns `Uint8ClampedArray` and `Extract.pixels` returns `Uint8Array`.

**Review by bigtimebuddy — COMMENTED** [review]
Can you fix the strict type issues?

**Pull request merged**